### PR TITLE
feat(log): adding log scale and max value filtering

### PR DESCRIPTION
**Pull request by eduelias**
Signed-off-by: Eduardo Elias Saleh <du7@msn.com>

Added Log Scaling for circle sizes, with option to choose which log function to use. Also, added a MaxValue, that enables the filtering of values > 10k ... > 100k, so the map has more meaning.

Tests updated to try on the new Scale.

### CODE REVIEW

File: src/data_formatter.ts [hunk 1]
```diff
@@ -7,9 +7,6 @@ export default class DataFormatter {
 
   setValues(data) {
     if (this.ctrl.series && this.ctrl.series.length > 0) {
-      let highestValue = 0;
```
Comment (PeterCsalaHbo) [inline src/data_formatter.ts:10]: Can I get some reasoning why did you remove the calculation of the `lowestValue`, `highestValue` and `valueRange` from this file?

File: src/partials/editor.html [hunk 2]
```diff
@@ -187,6 +187,16 @@ <h6 ng-show="ctrl.panel.locationData === 'table' || ctrl.panel.locationData ===
       <input type="text" class="input-small gf-form-input width-10" ng-model="ctrl.panel.esMetric" ng-change="ctrl.refresh()" ng-model-onblur
       />
     </div>
+    <div class="gf-form" ng-show="ctrl.panel.locationData === 'geohash'">
+      <gf-form-switch class="gf-form" ng-show="ctrl.panel.locationData === 'geohash'" label="Apply log Scale" label-class="width-12" checked="ctrl.panel.isLogScale" ng-model="ctrl.panel.isLogScale" on-change="ctrl.render()"></gf-form-switch>
```
Comment (PeterCsalaHbo) [inline src/partials/editor.html:191]: Why is it `label-class="width-12"` and not `with-10`?

As I can see most of the `input` and `gf-form-switch` are `width-10`

File: src/worldmap.test.ts [hunk 3]
```diff
@@ -387,4 +376,46 @@ describe('Worldmap', () => {
...
+      ctrl.panel.logFunction = 'log1p';
+      worldMap.drawCircles();
+    });
+
+    it('should draw three circles on the map', () => {
+      expect(worldMap.circles.length).toBe(3);
+    });
+
+    it('should create a circle with min circle size for smallest value size', () => {
+      expect(worldMap.circles[0].options.radius).toBe(2);
```
Comment (PeterCsalaHbo) [inline src/worldmap.test.ts]: How the `circles` are ordered?

I would suggest to use `find` and use the country code rather relying on indexing.

In my opinion this logic is fragile, because it assumes ordering.

File: src/worldmap.test.ts [hunk 4]
```diff
@@ -387,4 +376,46 @@ describe('Worldmap', () => {
...
+    });
+
+    it('should create a circle with circle size 9 for mid value size', () => {
+      expect(worldMap.circles[1].options.radius).toBe(9);
+    });
+
+    it('should create a circle with max circle size for largest value size', () => {
+      expect(worldMap.circles[2].options.radius).toBe(10);
```
Comment (PeterCsalaHbo) [inline src/worldmap.test.ts]: Would it make sense to create constants for `MinSize` and `MaxSize` and refer to them in the `beforeEach` and `it` methods?

File: src/worldmap.test.ts [hunk 5]
```diff
@@ -387,4 +376,46 @@ describe('Worldmap', () => {
     worldMap = new WorldMap(ctrl, document.getElementsByClassName('mapcontainer')[0]);
     worldMap.createMap();
   }
+
+  describe('when apply log is set', () => {
```
Comment (PeterCsalaHbo) [inline src/worldmap.test.ts:380]: Would it make sense to write test suites for other `logFunction`s as well?

Currently we are testing only `log1p`

File: src/worldmap_ctrl.ts [hunk 6]
```diff
@@ -35,6 +36,8 @@ const panelDefaults = {
   showLegend: true,
   mouseWheelZoom: false,
   hideLocationName: false,
+  isLogScale: false,
+  logFunction: 'log10',
```
Comment (PeterCsalaHbo) [inline src/worldmap_ctrl.ts]: Minor, but if the `isLogScale` is `false` by default then `logFunction` could be `undefined` by default (as well), if my understanding is correct.

File: src/worldmap.ts [hunk 7]
```diff
@@ -101,11 +104,20 @@ export default class WorldMap {
     return !_.isEqual(locations, dataPoints);
   }
 
+  // Determines if the log scale is applied and chooses which value to use
+  // Used when calculating circle sizes
+  getValue(dataPoint) {
+    return parseInt(this.ctrl.panel.isLogScale ? Math[this.ctrl.panel.logFunction](dataPoint.value) : dataPoint.value)
```
Comment (PeterCsalaHbo) [inline src/worldmap.ts]: This expression looks super ugly :)

I do believe if you split this into multiple lines it would greatly improve the legibility.

File: src/worldmap.ts [hunk 8]
```diff
@@ -140,6 +156,10 @@ export default class WorldMap {
   }
 
   updateCircles(data) {
+    const sizes = data.map(x => this.getValue(x));
+    this.lowestValue = Math.min(...sizes);
+    this.highestValue = Math.max(...sizes);
+    this.valueRange = this.highestValue - this.lowestValue;
```
Comment (PeterCsalaHbo) [inline src/worldmap.ts]: The exact same logic can be seen in the `createCircles` method.

Would it make sense to extract the logic into a method, which receives the `sizes` as a parameter?

File: src/worldmap.ts [hunk 9]
```diff
@@ -181,14 +201,14 @@ export default class WorldMap {
   }
 
   calcCircleSize(dataPointValue) {
-    const circleMinSize = parseInt(this.ctrl.panel.circleMinSize, 10) || 2;
+    let circleMinSize = parseInt(this.ctrl.panel.circleMinSize, 10) || 2;
```
Comment (PeterCsalaHbo) [inline src/worldmap.ts]: Why did you change it to `let`?

If there is a valid reasoning for that then shouldn't we have to change `circleMaxSize` as well?

File: src/worldmap.ts [hunk 10]
```diff
@@ -181,14 +201,14 @@ export default class WorldMap {
...
     const circleMaxSize = parseInt(this.ctrl.panel.circleMaxSize, 10) || 30;
 
-    if (this.ctrl.data.valueRange === 0) {
+    if (this.valueRange === 0) {
       return circleMaxSize;
     }
 
-    const dataFactor = (dataPointValue - this.ctrl.data.lowestValue) / this.ctrl.data.valueRange;
+    let dataFactor = (dataPointValue - this.lowestValue) / this.valueRange;
```
Comment (PeterCsalaHbo) [inline src/worldmap.ts]: Why is it `let`?

**Review by PeterCsalaHbo — CHANGES_REQUESTED** [review]

File: src/data_formatter.ts (hunk 1 above)
Comment (eduelias) [reply src/data_formatter.ts:10]: Well, this calculation was done when the data was loaded and, if we filtered on the "map side", those values would still be the same. You see, as they've built this plugin, there's two important "events", data load and screen load. Those calculations were happening on the data load (only when you load the scree or reload the data). Now, as we're changing how the circle's sizes are calculated with the data loaded, we would need to reload the data every time we changed the configurations. Now, this calculation is on the screen load(or reload). If the configuration changes, we calc it again (with the new values) and show them again.
Also, this formatter has a lot of duplicated and messy code. My intentions here are to clean up this file as best as I can, remove some duplications and make it more simple. Removing this from here helped this a lot.
If you look into Worldmap.ts:147, you gonna see how I did it waaaaaaaaaaay simpler, using Math.min and Math.max.

File: src/partials/editor.html (hunk 2 above)
Comment (eduelias) [reply src/partials/editor.html:191]: The labels are width-12 and the inputs width-10. The input "Hide local names" was, indeed, with the wrong width (10) on it's label, fixing now:
image

File: src/worldmap.test.ts (hunk 3 above)
Comment (eduelias) [reply src/worldmap.test.ts]: Ah, yes, this comes from a data builder, where we have control over the creation of the circles. I also believe that using find would be nice, but I just used the default from other tests.

File: src/worldmap.test.ts (hunk 4 above)
Comment (eduelias) [reply src/worldmap.test.ts]: There are tests where those max size and min sizes are different and, even, some tests (I wrote before) that uses their own internal constants. That one was just a copy from a previous test. But I'll create this const anyway ... ;)

File: src/worldmap.test.ts (hunk 5 above)
Comment (eduelias) [reply src/worldmap.test.ts:380]: Ouch, in fact that field I created to help me choose the best function to differentiate the circles better ... I wasn't going to commit the field altogether hehehe ... but, now that I did, writing tests for it ;)

File: src/worldmap_ctrl.ts (hunk 6 above)
Comment (eduelias) [reply src/worldmap_ctrl.ts]: I understand but I don't wanna do that. Why: Because, when you check the isLogScale, if you don't change the input, the log function would be undefined (requiring the user to actively choose a function on the selectbox). This way, the selectbox will display the default one, shown here and, if the user wants to change, than he can.
Otherwise I would need to check if it is set, use a default if not and, also, make sure the select shows the default if none selected. 10 lines of code when the solution above did it.

File: src/worldmap.ts (hunk 7 above)
Comment (eduelias) [reply src/worldmap.ts]: I DO agree! :)

File: src/worldmap.ts (hunk 9 above)
Comment (eduelias) [reply src/worldmap.ts]: This was something I did on the first iteration of this development, that I chose to undo because it was too ugly. Forgot to roll THAT back ... done! :D

File: src/worldmap.ts (hunk 10 above)
Comment (eduelias) [reply src/worldmap.ts]: Same as the let on circleMinSize... obsolete left behind.

File: src/worldmap.ts (hunk 8 above)
Comment (eduelias) [reply src/worldmap.ts]: awkward ...  did it even better hehehe

File: src/worldmap.test.ts (hunk 5 above)
Comment (eduelias) [reply src/worldmap.test.ts:380]: After a math study over the subject, just found out that, the way it is calculated, the base of log doesn't really matters after all. 
Look at the results here: https://www.desmos.com/calculator/1w0qzusjys
So, I AM removing this field.

File: src/worldmap.ts [hunk 11]
```diff
@@ -101,11 +104,31 @@ export default class WorldMap {
     return !_.isEqual(locations, dataPoints);
   }
 
+  // Determines if the log scale is applied and chooses which value to use
+  // Used when calculating circle sizes
+  getValue(dataPoint) {
```
Comment (PeterCsalaHbo) [inline src/worldmap.ts]: Please, please please we are using **Type**Script, not *JavasScript*.

Please specify the input parameter's **type** and the return **type** for the newly introduced methods.

**Review by PeterCsalaHbo — DISMISSED** [review]

**Review by PeterCsalaHbo — APPROVED** [review]
